packages/grafana-data/src/transformations/transformers/fillGaps.ts

```diff
@@ -0,0 +1,61 @@
+import { MutableDataFrame, DataFrame, FieldType } from '../..';
+
+import { map } from 'rxjs/operators';
+
+import { DataTransformerID } from './ids';
+import { DataTransformerInfo } from '../../types/transformations';
+
+export interface FillGapsTransformerOptions {
+  interval?: number;
+  fillWith: any;
+}
+
+export const fillGapsTransformer: DataTransformerInfo<FillGapsTransformerOptions> = {
+  id: DataTransformerID.fillGaps,
+  name: 'Fill Gaps',
+  description: '',
+  defaultOptions: {
+    fillWith: null,
+  },
+  operator: (options) => (source) =>
+    source.pipe(
+      map((dataFrames) => {
+        if (!Array.isArray(dataFrames)) {
+          return dataFrames;
+        }
+        return dataFrames.map((f) => fillGaps(f, options));
+      })
+    ),
+};
+
+const getInterval = (frame: DataFrame) => frame.fields.map((f) => f.config.interval ?? 0).find((i) => i > 0) ?? 0;
+
+export const fillGaps = (df: DataFrame, options?: FillGapsTransformerOptions): DataFrame => {
+  const frame = new MutableDataFrame(df);
+  const interval = options?.interval ?? getInterval(frame);
+  if (interval === 0) {
+    return frame;
+  }
+
+  const timeField = frame.fields.find((f) => f.type === FieldType.time);
+  if (!timeField) {
+    return frame;
+  }
+  const valueFields = frame.fields.filter((f) => f !== timeField);
+
+  let rowIdx = 0;
+  let start = timeField.values.get(0);
+  for (const end of timeField.values.toArray()) {
+    for (let t = start; t < end; t += interval) {
+      timeField.values.insert(rowIdx, t);
+      valueFields.forEach((f) => f.values.insert(rowIdx, options?.fillWith ?? null));
```

Comment:
just use the raw array and no need to change vector interface

Reply:
updated in 334ffa3